### PR TITLE
Refactoring ScalarMatcher to be more friendly and tests only true scalar types

### PR DESCRIPTION
**Pull request by diegoholiveira**
I rewrote the ScalarMatchers to be more like ObjectStateMatcher, and now the matcher it's more friendly:
- `shouldBeInt`
- `shouldBeFloat`
- `shouldBeString`
- `shouldBeBoolean`

Also, the matcher now verifies only true scalar types.

### CODE REVIEW

**Comment by everzet** [comment]
If you'll leave it like that, this matcher will interfere with `ObjectStateMatcher` - `ObjectStateMatcher` will never be called as this matcher will always throw `FunctionNotFoundException` instead of leaving stuff for other matchers to handle. `ObjectStateMatcher` throw `MethodNotFoundException`, because it has lowest priority and proven to be called last (when no other matcher supports message).

Also, we're not talking about functions here, we're providing OOP tool, so this exception is useless in my opinion.

I think you should remove `FunctionNotFoundException` from this PR and instead move function existence check into `support(...)` - matcher just shouldn't be evaluated if there's no checker function defined.

File: src/PHPSpec2/Matcher/ScalarMatcher.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+     * @return boolean
+     */
+    public function positiveMatch($name, $subject, array $arguments)
+    {
+        preg_match($this->regex, $name, $matches);
+
+        $expected = strtolower($matches[2]);
+
+        switch ($expected) {
+
```
Comment (everzet) [inline src/PHPSpec2/Matcher/ScalarMatcher.php]: Useless unconventional space.

File: src/PHPSpec2/Matcher/ScalarMatcher.php [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            case 'boolean':
+                $func = 'is_bool';
+                break;
+
```
Comment (everzet) [inline src/PHPSpec2/Matcher/ScalarMatcher.php]: Same.

**Comment by MarcelloDuarte** [comment]
@everzet are you following up on @diegoholiveira fixes? Is the priority fixed? Do you want me to look, or can you?

**Comment by everzet** [comment]
@MarcelloDuarte if you have time, it'd be better for you too look into it. :+1:

**Comment by diegoholiveira** [comment]
So guys, there any change required to merge this Matcher?

**Comment by MarcelloDuarte** [comment]
Hi @diegoholiveira, I am looking at this now. I wrote a spec:

```
<?php
class Test extends ObjectBehavior
{
    function it_should_return_some_string()
    {
        $this->gimeSomeString()->shouldBeString();
    }
}
```

I get this:

```
9  ! it should be initializable
      matcher beString not found. Have you registered it properly? 
```

Because I am returning a resource:

```
    public function gimeSomeString()
    {
        return STDOUT;
    }
```

I think you are nearly there by the way. You did a great job understanding how the matchers work.

**Comment by diegoholiveira** [comment]
Thanks dude. @everzet help's me a lot with this PR :)

This behaviour that you describe is the rigth one, rigth?
This happens because the "is_scalar($subject)" on the method supports.

**Comment by MarcelloDuarte** [comment]
Yes. We expect $subject to be scalar, but it may not be, so the message needs to tell us that it is not. `$this->gimeSomeString()->shouldBeString();` is failing, but the message is not what the developer needs to see. You want something similar to what you have done with the other scalars: expected `string`, but got `resource`.

**Comment by diegoholiveira** [comment]
@MarcelloDuarte fixed.

I also merge phpspec2/master into my branch to be more easy to merge my PR :)

**Pull request merged**